runtime/src/accounts_db.rs

```diff
@@ -5119,6 +5133,8 @@ impl AccountsDb {
                 );
                 let count = store.remove_account(account_info.stored_size, reset_accounts);
                 if count == 0 {
+                    self.dirty_stores
+                        .insert((*slot, store.append_vec_id()), store.clone());
```

Comment:
Another annoying case to think about is that right now unrooted slots can also prevent zero lamport cleanup due to them counting toward the `ref_count_from_storage`, but only rooted slots are included in `account_infos.len()` here:  https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L1483. So for example if we had:

```
<Slot S, Rooted>                   <Slot S+1, Unrooted>               <Slot S+2, Rooted>
Y=1                                               Y=2, Z=1                                         Y=0
```

The `ref_count_from_storage` of `Y` is `2`: https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L1812, because even though *rooted* slot `S` is dead, *unrooted* slot `S+1` is alive and adds to the ref count for `Y`. Meanwhile `account_infos` only contains the *rooted* update in slot `S+2`, so this check fails https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L1483.

Luckily, later, when the bank for slot `S+1` is dropped, cleanup happens in `purge_slots()` and calls into `purge_slot_storage()`: https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L3394, which should should ad slot `S+1` to the dirty stores: https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L5136-L5138. 

I think it would be good to add a test here. I think you can copy and paste this one: https://github.com/carllin/solana/commit/c2683cdc983a7a574047d45dbba183b8a4f60009#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR11025-R11077, which tests exactly this case 😃 .

Reply:
This test fails on master as well. I think we can do another change to handle this case. It doesn't seem to be related to this change specifically.

Reply:
> This test fails on master as well

Shoot, that's because the test was wrong! 

> It doesn't seem to be related to this change specifically

Yeah it's not directly related, but the removal of the zero-lamport accounts in the candidate key generation had a chance of making the clean fail here: https://github.com/carllin/solana/blob/5bb5284438f0c5a545a4db53924f28e5759a3591/runtime/src/accounts_db.rs#L11670-L11673 if the slot storage wasn't being correctly added to the `dirty_stores` list in the `purge_slots()` path, so it was just an additional corner case to consider.

I extracted the test into a separate PR, which passes now on master and with these changes, so not a blocker here 😃

Reply:
Awesome, thanks!